### ISSUE TITLE
Unexpected argument passed when "math-calling" functions

### ISSUE THREAD

**Issue by choking-chips** [typst/typst#2993]
### Description

Hi! I'm writing a document with a function which makes its content having the following appearance:
- with space on both sides
- with an underline
- bold font

So I let a function like this:
```typst
#let blank(content) = {
  box(
    inset: (left: 1em, right: 1em, bottom: 2pt),
    baseline: 2pt,
    stroke: (bottom: black),
    strong(content)
  )
}
```
It was good enough, until I use it in a math field, like:
```typst
$(101000)_2=(blank(40))_10$
```
I was expecting the following result:
![image](https://github.com/typst/typst/assets/52846426/b5fc793f-c72e-4508-924a-2d69f2815c7f)

But unfortunately, it actually showed with no bold style:
![image](https://github.com/typst/typst/assets/52846426/1ca2faed-6398-4c4e-baf8-9fae66aa4873)

It showed correctly when you write as follows:
```typst
$(101000)_2=(#blank($40$))_10$
```
```typst
$(101000)_2=(blank(4 0))_10$
```
```typst
$(101000)_2=(blank(equation(40)))_10$
```

But, all the solutions looked weird. The easier, the better, isn't it?

Then I tried to investigate how it caused. What I find in short:
- If the argument is just a number, it will be passed as a `text`, though it is hardly distinguishable from simple texts and cannot be made bold using `strong()` like simple texts.
- Otherwise, it will probably be passed as a `sequence`, which we can filter out and simply use `math.bold()`.

I'm wondering how can this be solved. Anyway, thank you for your wonderful project!

### Operating system

Web app, Windows

### Typst version

- [X] I am using the latest version of Typst

**Comment by Enivex** [comment]
There are some weird interactions going on here, but text markup functions shouldn't really affect math at all

https://github.com/typst/typst/issues/1155

**Comment by choking-chips** [comment]
> There are some weird interactions going on here, but text markup functions shouldn't really affect math at all

Yeah, I think my words are too much, maybe I need to fix my description, sorry.

The point is that there is no way for my function to check whether the argument is a `math` or not. At least I can't make it at present.

**Comment by choking-chips** [comment]
To make it clearer, I'll show an example.

```typst
#let blank(content) = {
  box(
    inset: (left: 1em, right: 1em, bottom: 2pt),
    baseline: 2pt,
    stroke: (bottom: black),
    [#content.func()]
  )
}
```
```typst
$(101000)_2=(blank(40))_10$
$(63)_10=(blank(3 upright(E)))_16$
```
![image](https://github.com/typst/typst/assets/52846426/f54188f3-36c1-4c44-9a1b-bf690951b1bb)
